### PR TITLE
 Improve gemini style guide

### PR DESCRIPTION
**Pull request by laxmareddyp**
## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: .gemini/styleguide.md [hunk 1]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+### Step 1: Identify the Model Architecture
+
+Before reviewing any file, determine the model's architecture type from its directory structure:
+
+- **Decoder-only LM** (e.g., Gemma, Llama, GPT2): Tokenizer → Embedding → Transformer Decoder → LM Head
+- **Multimodal LM** (e.g., Gemma3, PaliGemma, Qwen3.5): VisionEncoder → Token Injection/Fusion → Transformer → LM Head
+- **Encoder-only** (e.g., BERT, RoBERTa): Tokenizer → Transformer Encoder → Task Head
+- **Encoder-Decoder** (e.g., T5, BART, Whisper): Tokenizer → Encoder → Cross-Attention Decoder → Task Head
+- **Vision model** (e.g., ViT, ResNet): ImageConverter → Vision Backbone → Classifier Head
+- **Multimodal + Audio** (e.g., Gemma3n): AudioConverter + VisionEncoder → Fusion → Transformer → LM Head
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md]: 3N model is not yet merged, it will not pick reference.
You can include Gemma4, which has multi model, audio, and video capabilities.

File: .gemini/styleguide.md [hunk 2]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+This is the most important section of this style guide. **You are an architecture-aware code reviewer for deep learning model implementations. You MUST NOT review any file in isolation. You MUST reason about the full model architecture, data flow, and cross-file contracts before commenting.**
+
+---
+
+### Step 1: Identify the Model Architecture
+
+Before reviewing any file, determine the model's architecture type from its directory structure:
+
+- **Decoder-only LM** (e.g., Gemma, Llama, GPT2): Tokenizer → Embedding → Transformer Decoder → LM Head
+- **Multimodal LM** (e.g., Gemma3, PaliGemma, Qwen3.5): VisionEncoder → Token Injection/Fusion → Transformer → LM Head
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md]: Same Qwen3.5 is not yet merged, let's use the examples and models we have.

File: .gemini/styleguide.md [hunk 3]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+
+- **`hidden_dim`** must match across embedding, decoder layers, attention, and MLP
+- **`num_heads` / `num_key_value_heads`** must be consistent between backbone config and attention layers (GQA ratio must be correct)
+- **`head_dim`** must equal `hidden_dim // num_heads` (or be explicitly configured)
+- **`intermediate_dim`** must match between backbone config and MLP layers
+- **`vocabulary_size`** must match between tokenizer, embedding layer, and LM head
+- **`max_sequence_length`** must be consistent between preprocessor packing and RoPE/position embeddings
+- **Vision/text embedding dimensions must align** at the fusion point in multimodal models
+- **Special token IDs** (image_token_id, audio_token_id, pad_token_id) must match between tokenizer config and preprocessor logic
+- **`rope_max_wavelength`** and **`rope_scaling_factor`** must match between backbone config and attention layers
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md]: These are not generic for all models right?

File: .gemini/styleguide.md [hunk 4]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+---
+
+### Strict Rules
+
+- **DO NOT give generic feedback** — every comment must reference the specific architectural context
+- **DO NOT review style unless it affects correctness** — focus on architecture, not formatting
+- **DO NOT flag patterns that exist in other established models** — if Gemma, Llama, or other models use the same pattern, it is intentional
+- **DO NOT assume missing context** — infer the design intent from the architecture and cross-file dependencies
+- **PRIORITIZE system-level correctness over local correctness** — a function that looks wrong in isolation may be correct in the context of the full model
+- **If you cannot explain why a pattern is wrong after understanding the full architecture, do not flag it**
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md]: This may make model not to give review on certain things, we better ignore and justify the false review from the model rather than not getting few review with these rules.

File: .gemini/styleguide.md [hunk 5]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+- Verify attention mask shapes match between preprocessor, backbone, and decoder
+- For multimodal models: verify vision embedding shapes align with text embedding shapes at the fusion point
+
+#### 5.4 Multimodal Correctness (if applicable)
+- Are modality placeholder tokens (image, audio) inserted and replaced correctly?
+- Do vision/audio indices correctly map to positions in the token sequence?
+- Is the fusion logic (embedding replacement, cross-attention, or concatenation) implemented correctly?
+- Are dummy inputs for missing modalities the correct shape?
+
+#### 5.5 Silent Failure Risks (CRITICAL)
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md:887]: Add a point to avoid .get() and default values in converter files.

File: .gemini/styleguide.md [hunk 6]
```diff
@@ -765,3 +765,146 @@ class TestMyModelConverter(TestCase):
...
+
+---
+
+### Step 5: Review Checklist
+
+When reviewing, you MUST perform ALL of the following checks. **Do not give generic feedback. Every comment must be specific and grounded in the architecture.**
+
+#### 5.1 Architecture Consistency
+- Does this file align with the overall model architecture?
+- Do the layer dimensions, head counts, and vocabulary sizes match the backbone config?
```
Comment (sachinprasadhs) [inline .gemini/styleguide.md]: What is this point about? Is it from original HF config?

**Comment by laxmareddyp** [comment]
@sachinprasadhs Thanks for the detailed review. Will look into it. 
@divyashreepathihalli Do you have any suggessions apart from sachin's comments, will push all together.

File: .gemini/styleguide.md [hunk 7]
```diff
@@ -11,7 +11,7 @@ When performing code reviews on pull requests, you must strictly adhere to the f
 
 5. **Respect Existing Repo Patterns**: Before suggesting review comments (like asking users to add boilerplate or specific patterns), actively check for existing design patterns across the repository. Do not suggest adding useless code or structures that contradict or fall outside the established Keras repo coding style.
 
-
+6. **Understand the Full Model Architecture Before Commenting**: KerasHub model implementations span multiple tightly coupled files (backbone, decoder, attention, layers, preprocessor, task, converter). **Do not isolate a single file or code snippet and comment on it without understanding how it fits into the overall architecture.** Before leaving any review comment, read all files in the model directory to understand the data flow, cross-file dependencies, and design intent. A pattern that looks unusual in isolation is often intentional when seen in the context of the full model. If you are unsure whether a pattern is correct, check if similar patterns exist in other established models in the repo before flagging it.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: remove this line " A pattern that looks unusual in isolation is often intentional when seen in the context of the full model." - which may not always be true.

File: .gemini/styleguide.md [hunk 8]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
 - Include comprehensive examples
 - Document edge cases and limitations
 - Keep documentation up-to-date with code changes
+
+---
+
+## Context-Aware Model Review
+
+This is the most important section of this style guide. **You are an architecture-aware code reviewer for deep learning model implementations. You MUST NOT review any file in isolation. You MUST reason about the full model architecture, data flow, and cross-file contracts before commenting.**
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: remove - You MUST NOT review any file in isolation

File: .gemini/styleguide.md [hunk 9]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+
+---
+
+### Step 1: Identify the Model Architecture
+
+Before reviewing any file, determine the model's architecture type from its directory structure:
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: determine -> understand

File: .gemini/styleguide.md [hunk 10]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+For every file you review, determine:
+
+- **What component is this?** (backbone, decoder, attention, layers, preprocessor, task, converter, tokenizer, test)
+- **Who calls this?** (e.g., backbone is called by the task model; decoder is called by backbone)
+- **What does this depend on?** (e.g., attention depends on RoPE config from backbone; preprocessor depends on tokenizer special tokens)
+
+KerasHub models have a standard component dependency:
+
+```
+Task (CausalLM, Classifier, etc.)
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: this structure is not exhaustive

File: .gemini/styleguide.md [hunk 11]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+        ├── Tokenizer
+        ├── ImageConverter (multimodal only)
+        └── AudioConverter (audio models only)
+```
+
+---
+
+### Step 3: Validate Architecture Invariants
+
+Config values must be consistent across all files in a model. Before flagging a mismatch, check that the values used in each component (backbone, decoder, attention, preprocessor, tokenizer) are consistent with what the backbone config defines.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: config values are not in the code

File: .gemini/styleguide.md [hunk 12]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+
+Common things to verify (where applicable to the model type):
+
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md:842]: not in the code base

File: .gemini/styleguide.md [hunk 13]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+- Sequence length limits are consistent between preprocessor and positional encoding
+- Special token IDs used in the preprocessor match those defined in the tokenizer
+- For multimodal models: embedding dimensions align at modality fusion points
+
+---
+
+### Step 4: Trace the Execution Flow
+
+Follow the data flow end-to-end before commenting on any individual component:
+
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md:855]: make sure the flow is completely exhaustive

File: .gemini/styleguide.md [hunk 14]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+### Step 5: Review Checklist
+
+When reviewing, you MUST perform ALL of the following checks. **Do not give generic feedback. Every comment must be specific and grounded in the architecture.**
+
+#### 5.1 Architecture Consistency
+- Does this file align with the overall model architecture?
+- Are the layer names consistent with how they are accessed from other files (e.g., `backbone.get_layer("transformer_layer_0")`)?
+
+#### 5.2 Cross-File Contract Validation
+- Are the inputs/outputs of this component compatible with the components that call it and that it depends on?
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: *of a component

File: .gemini/styleguide.md [hunk 15]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+- The input names used in `keras.Input(name=...)` must match the dictionary keys produced by the preprocessor's `call()` method
+- The functional graph's output shape must match what the task model (e.g., `CausalLM`) expects when it accesses `self.backbone(...)` outputs
+
+---
+
+## Mandatory Pre-Review Checklist
+
+Before leaving ANY review comment, you MUST:
+
+1. **Search for precedent first**: Search for the same pattern in at least 3 other model directories in `keras_hub/src/models/`. If the pattern exists in multiple merged models, it is an established convention — do not flag it. The codebase has 75+ models; new models should follow existing patterns, not invent new ones.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: " new models should follow existing patterns, not invent new ones.:" - not always true

File: .gemini/styleguide.md [hunk 16]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+
+2. **Check the parent model**: If the PR adds a variant model (e.g., MoE version of a dense model), compare every flagged pattern against the parent model first. If the parent uses the same code, the variant should too.
+
+3. **Verify your suggested fix compiles**: If you suggest a code change, verify the Python syntax is correct. Do not suggest ternary expressions, import restructuring, or API calls without confirming they work.
+
+4. **Do not flag the same issue twice**: If you identify a pattern concern, flag it once with a single comment. Do not repeat the same suggestion on multiple instances in the same PR.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: remove this - 4. **Do not flag the same issue twice**: If you identify a pattern concern, flag it once with a single comment. Do not repeat the same suggestion on multiple instances in the same PR.

File: .gemini/styleguide.md [hunk 17]
```diff
@@ -765,3 +765,241 @@ class TestMyModelConverter(TestCase):
...
+
+5. **Distinguish bugs from style preferences**: Only flag issues that would cause incorrect behavior, data loss, or test failures. Do not flag working code that could theoretically be written differently unless it violates an explicit rule in this guide.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md]: remove 5 as well - 5. **Distinguish bugs from style preferences**: Only flag issues that would cause incorrect behavior, data loss, or test failures. Do not flag working code that could theoretically be written differently unless it violates an explicit rule in this guide.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

File: .gemini/styleguide.md (hunk 7 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: Removed

File: .gemini/styleguide.md (hunk 9 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: Updated

File: .gemini/styleguide.md (hunk 10 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: Added more now.

File: .gemini/styleguide.md (hunk 11 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: Updated

File: .gemini/styleguide.md (hunk 12 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md:842]: updated

File: .gemini/styleguide.md (hunk 13 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md:855]: Added more now.

File: .gemini/styleguide.md (hunk 14 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: updated

File: .gemini/styleguide.md (hunk 15 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: removed this line

File: .gemini/styleguide.md (hunk 16 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: removed

File: .gemini/styleguide.md (hunk 17 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: removed

File: .gemini/styleguide.md (hunk 8 above)
Comment (laxmareddyp) [reply .gemini/styleguide.md]: updated

File: .gemini/styleguide.md [hunk 18]
```diff
@@ -990,16 +1053,11 @@ KerasHub backbones use `keras.Input()` to build a Keras Functional graph in `__i
...
-
-2. **Check the parent model**: If the PR adds a variant model (e.g., MoE version of a dense model), compare every flagged pattern against the parent model first. If the parent uses the same code, the variant should too.
-
-3. **Verify your suggested fix compiles**: If you suggest a code change, verify the Python syntax is correct. Do not suggest ternary expressions, import restructuring, or API calls without confirming they work.
+## Review Anti-Patterns — DO NOT Flag These
 
-4. **Do not flag the same issue twice**: If you identify a pattern concern, flag it once with a single comment. Do not repeat the same suggestion on multiple instances in the same PR.
+Before leaving a review comment, you MUST check whether the pattern already exists in at least 2 other merged models in the repository. If it does, the pattern is an established convention and MUST NOT be flagged.
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md:1058]: Gemini need not do this for every comment. It does not really apply for every comment it leaves. This is too specific. Its okay to get a Gemini comment that you can ignore rather than guiding it too narrowly.

File: .gemini/styleguide.md [hunk 19]
```diff
@@ -990,16 +1053,11 @@ KerasHub backbones use `keras.Input()` to build a Keras Functional graph in `__i
...
 
-5. **Distinguish bugs from style preferences**: Only flag issues that would cause incorrect behavior, data loss, or test failures. Do not flag working code that could theoretically be written differently unless it violates an explicit rule in this guide.
+### Before Commenting, Always:
```
Comment (divyashreepathihalli) [inline .gemini/styleguide.md:1060]: no
